### PR TITLE
Do not send loop counters to device

### PR DESCRIPTION
**Pull request by josevalim**

### CODE REVIEW

**Review by polvalente — APPROVED** [review]

**Comment by josevalim** [comment]
The CI failure is a precision error. The ONNX is also failing on main. :)

File: lib/axon/loop.ex [hunk 1]
```diff
@@ -1065,10 +1061,7 @@ defmodule Axon.Loop do
               {:halt, {:halt_loop, state}}
 
             {:continue, state} ->
-              iters = Nx.to_number(iters)
-              max_iters = Nx.to_number(max_iters)
-
-              if iters > max_iters and max_iters != -1 do
+              if iters > max_iters and max_iters != nil do
```
Comment (josevalim) [inline lib/axon/loop.ex:1064]:
```suggestion
              if iters > max_iters and max_iters != -1 do
```

**Review by seanmor5 — APPROVED** [review]
Thank you :)

File: lib/axon/loop.ex [hunk 2]
```diff
@@ -1052,23 +1044,32 @@ defmodule Axon.Loop do
...
+
+          {new_step_state, new_metrics} =
+            maybe_jit(batch_fn, [data, iters, step_state, metrics], jit_compile?, jit_opts)
+
+          state = %{
+            state
+            | iteration: iters + 1,
+              step_state: new_step_state,
+              metrics: new_metrics
+          }
```
Comment (josevalim) [inline lib/axon/loop.ex:1062]: @seanmor5 now it all works! I found out the simplest approach is actually to not pass the state into defn/jit. Only the parts that we need and then get them back. This way there is no confusion about the boundaries and about what goes and what comes back.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**